src/dbnode/client/circuitbreaker/middleware/middleware.go

```diff
@@ -0,0 +1,245 @@
+package middleware
+
+import (
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/circuitbreakererror"
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/internal/circuitbreaker"
+	"github.com/m3db/m3/src/dbnode/generated/thrift/rpc"
+	"github.com/uber-go/tally"
+	tchannel "github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+	"go.uber.org/zap"
+	"sync/atomic"
+)
+
+// circuitBreakerClient is a client that wraps a TChannel client with a circuit breaker.
+type circuitBreakerClient struct {
+	enabled    bool
+	shadowMode bool
+	logger     *zap.Logger
+	cb         *atomic.Value // *circuitbreaker.Circuit
+	metrics    *circuitBreakerMetrics
+	host       string
+	next       rpc.TChanNode
+}
+
+// m3dbMiddleware is a function that takes a TChannel client and returns a circuit breaker client interface.
+type m3dbMiddleware func(rpc.TChanNode) CircuitBreakerClient
+
+// CircuitBreakerClient defines the interface for a circuit breaker client.
+type CircuitBreakerClient interface {
+	rpc.TChanNode
+}
+
+// NewCircuitBreakerMiddleware creates a new circuit breaker middleware.
+func NewCircuitBreakerMiddleware(config Config, logger *zap.Logger, scope tally.Scope, host string) m3dbMiddleware {
+	c, err := circuitbreaker.NewCircuit(config.CircuitBreakerConfig)
+	if err != nil {
+		logger.Warn("failed to create circuit breaker", zap.Error(err))
+	}
+	cb := &atomic.Value{}
+	cb.Store(c)
+
+	return func(next rpc.TChanNode) CircuitBreakerClient {
+		return &circuitBreakerClient{
+			enabled:    config.Enabled,
+			shadowMode: config.ShadowMode,
+			next:       next,
+			logger:     logger,
+			host:       host,
+			metrics:    newMetrics(scope, host),
+			cb:         cb,
+		}
+	}
+}
+
+// withBreaker executes the given call with a circuit breaker if enabled.
+func withBreaker[T any](c *circuitBreakerClient, ctx tchannel.ContextWithHeaders, call func() error) error {
+	if !c.enabled {
+		return c.executeWithoutBreaker(call)
+	}
+
+	cb := c.getCircuit()
+	if cb == nil || !cb.IsRequestAllowed() {
+		return c.handleRejectedRequest()
+	}
+
+	return c.executeWithBreaker(cb, call)
+}
+
+// executeWithoutBreaker executes the given call without a circuit breaker.
+func (c *circuitBreakerClient) executeWithoutBreaker(call func() error) error {
+	c.logger.Debug("circuit breaker disabled, calling next", zap.String("host", c.host))
+	return call()
+}
+
+// getCircuit retrieves the circuit breaker from the atomic value.
+func (c *circuitBreakerClient) getCircuit() *circuitbreaker.Circuit {
```

Comment:
nit: go tends to prefer not prefixing getters with `Get`: https://go.dev/doc/effective_go#Getters